ch4/gpu: Add GPU RDMA support for RMA

abrooks98 opened the pull request:
## Pull Request Description

This PR adds GPU RDMA support for RMA.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

abrooks98 commented:
test:mpich/ch4/most
test:mpich/ch4/gpu

abrooks98 commented:
test:mpich/ch4/most
test:mpich/ch4/gpu

hzhou commented on src/mpl/include/mpl_gpu.h:
```diff
@@ -95,6 +95,11 @@ static inline int MPL_gpu_query_pointer_is_dev(const void *ptr, MPL_pointer_attr
+{
+    return 0;
+}
+
```
Is this supposed to be in `mpl_gpu_fallback.c`? The inline function here would conflict with the actual definition, right?

abrooks98 replied:
Since this is in the `#ifndef MPL_HAVE_GPU` block I think it is effectively the same as being defined in `mpl_gpu_fallback.c`, but I can move it there anyways

abrooks98 replied:
Done

hzhou reviewed (commented):
Please add commit messages to the first two commits. The first commit need explain the purpose of query strict device, e.g. what does "strict device" mean? The second commit need explain the changes. It potentially needs split since it touches both the pt2pt path and rma path.

abrooks98 commented:
> Please add commit messages to the first two commits. The first commit need explain the purpose of query strict device, e.g. what does "strict device" mean? The second commit need explain the changes. It potentially needs split since it touches both the pt2pt path and rma path.

Added descriptions to the commits and also split the second commit into two

abrooks98 commented:
test:mpich/ch4/most
test:mpich/ch4/gpu

abrooks98 commented:
Tests are clean, ready for more reviews

raffenet reviewed (approved):
LGTM

hzhou commented on src/include/mpir_gpu.h:
```diff
@@ -105,6 +105,15 @@ MPL_STATIC_INLINE_PREFIX bool MPIR_GPU_query_pointer_is_dev(const void *ptr)
+
+    return false;
+}
+
```
Should this be removed?

abrooks98 replied:
I think we need to keep this because it is called during `win_allgather` where there are no guards for whether or not GPUs are enabled. `MPL_gpu_query_pointer_is_strict_dev` would have a hard failure for ZE if it was called without calling `MPL_gpu_init` first

hzhou replied:
Oh, my mistake. I thought this was in the MPL header.

hzhou replied:
Okay, the next question is why not all paths call this function?

abrooks98 replied:
That I will fix :)

abrooks98 replied:
Done.

Seems we might need to do a similar cleanup for calls to `MPL_gpu_query_pointer_is_dev`; that can be done in a separate PR

hzhou reviewed (approved):
LGTM

hzhou commented:
Let's run tests one more time and we can merge after that.

test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi

abrooks98 commented:
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
